--- a/src/Bulk.php
+++ b/src/Bulk.php
@@ -131,6 +131,10 @@ class Bulk
      */
     public function addDocument(Document $document, ?string $opType = null): self
     {
+        if (!$document->hasRetryOnConflict()) {
+            $retry = $this->_client->getConfigValue('retryOnConflict', 0);
+
+            if ($retry > 0) {
+                $document->setRetryOnConflict($retry);
+            }
+        }
+
         $action = AbstractDocumentAction::create($document, $opType);
 
         return $this->addAction($action);
@@ -155,6 +159,10 @@ class Bulk
      */
     public function addScript(AbstractScript $script, ?string $opType = null): self
     {
+        if (!$script->hasRetryOnConflict()) {
+            $retry = $this->_client->getConfigValue('retryOnConflict', 0);
+
+            if ($retry > 0) {
+                $script->setRetryOnConflict($retry);
+            }
+        }
+
         $action = AbstractDocumentAction::create($script, $opType);
 
         return $this->addAction($action);